### ISSUE TITLE
Number of threads limited to number of cores (omp_get_num_procs())

### ISSUE THREAD

**Issue by simon-lenau** [Rdatatable/data.table#5888]
I am using data.table on a compute infrastructure with 2x AMD Epyc 7773x - (128 Cores, 256 Threads).

Although `OMP_THREAD_LIMIT` and `omp_get_thread_limit()` return the correct number of threads, data.table currently uses only two threads because `omp_get_num_procs()` returns `2`.

I am aware that this behavior is documented in the help files, but it hardly seems a desirable behavior for me. 
A way to circumvent this issue (e.g. with environment variables and warnings) without having to customize data.table's `openmp-utils.c` file would be helpful.

**Comment by jangorecki** [comment]
From what you write it sounds like it is abnormal that omp_get_num_procs() returns 2. Is that correct? Why it returns 2 then?

DT should use 50% of threads by default.
What's your openmp version? Compiler version? OS?
See `getDTthreads(verbose=TRUE)`

**Comment by simon-lenau** [comment]
To be honest, I am not sure why it returns 2 and whether that's _correct_. Currently, all I can say is that it indeed returns 2.

``` 
> getDTthreads(verbose=TRUE)
  OpenMP version (_OPENMP)       201511
  omp_get_num_procs()            2
  R_DATATABLE_NUM_PROCS_PERCENT  unset (default 50)
  R_DATATABLE_NUM_THREADS        unset
  R_DATATABLE_THROTTLE           unset (default 1024)
  omp_get_thread_limit()         2147483647
  omp_get_max_threads()          42
  OMP_THREAD_LIMIT               unset
  OMP_NUM_THREADS                42
  RestoreAfterFork               true
  data.table is using 2 threads with throttle==1024. See ?setDTthreads.
```

Afaik this corresponds to OpenMP 4.5.

I am using the latest (v4.3.2) [docker image of r-base](https://hub.docker.com/_/r-base) for testing this behaviour.

Compiler is
```
> gcc --version
gcc (Debian 13.2.0-7) 13.2.0
```

The OS is debian:
```
> cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux trixie/sid"
NAME="Debian GNU/Linux"
VERSION_CODENAME=trixie
```

**Comment by jangorecki** [comment]
For my machine, having 4 cores, 8 threads, `omp_get_num_procs()` returns 8.
This must be the ultimate reason why you only have 2 threads available. 
According to spec

> The omp_get_num_procs routine returns the number of processors available to the device.

Therefore, despite having so many cores, it seems that only 2 are actually available. This may be a deliberate setting by an admin of the machine.

I would suggest to narrow down the problem to C code only, by writing C program that simply prints `omp_get_num_procs()`. Compile it from gcc, and after confirming it prints `2`, reach out to your admin, asking why is it so.

**Comment by simon-lenau** [comment]
Thanks! Indeed, `omp_get_num_procs()` also returns 2 if I use it in C++ directly.
My wild guess was that this represents the two Epyc processors rather than their cores.
By the way, the number of detected threads (e.g.) via `parallel` is correct (256):
```
> parallel::detectCores()
[1] 256
```
Anyway, I will check with the admin and see whether there's something wrong with OpenMP!

**Comment by tdhock** [comment]
it seems like a bug in omp_get_num_procs, so I would suggest submitting this issue to openmp devs.

**Comment by jangorecki** [comment]
There are no single openmp devs. Different compilers uses different implementations, implemented by different devs. I believe it is a matter of setup of openmp by an admin of the machine.

**Comment by eddelbuettel** [comment]
(rocker/r-base maintainer here, also scratching his head)  Could Docker be limiting your 'view' of the machine? This sounds ... weird.  I get 12 (correct value on my not exactly new i7-8700) either way:

```sh
edd@rob:~$ Rscript -e 'parallel::detectCores()'
[1] 12
edd@rob:~$ docker run --rm -ti rocker/r-base Rscript -e 'parallel::detectCores()'
[1] 12
edd@rob:~$ 
```

**Comment by ben-schwen** [comment]
Since you are using docker, I would guess your admin limits the number of cores.

Can you run a container with `docker run --cpuset-cpus="0-3" --rm -it --entrypoint bash r-base` and post the output of `nproc`. It should be 4.

In this case `parallel::detectCores()` also outputs the number of cores and not the number of cpus available in the docker image.

**Comment by eddelbuettel** [comment]
Occam's razor: You likely have one of the environment variables set that OpenMP respects and so you are getting what you should be getting.  Try `Sys.getenv("OMP_THREAD_LIMIT")`  and/or `Sys.getenv()` and grep in what comes back for `OMP`.

**Comment by simon-lenau** [comment]
Thanks @ben-schwen and @eddelbuettel for your answers!
I had already set `OMP_THREAD_LIMIT`, but data.table does not use this limit if the result of `omp_get_num_procs()` is lower. 

With the help of our administrator, I found out that the problem is related to how our scheduler assigns cores to containers. It turned out that `omp_get_num_procs()` was correct - and so was the data.table with relying on it!

(It turned out that `parallel::detectCores()` incorrectly returned the number of cores I was expecting. 
I think this is because it use `/proc/cpuinfo`, which seems to be mounted from the host system).
